### PR TITLE
Modular Content Widget 

### LINKED ISSUE #1066: Flexible Layouts Custom Widget
I have an Idea for a custom widget. It is based mostly on Advanced Custom Fields "[Flexible Field](https://www.advancedcustomfields.com/resources/flexible-content/)" which allows for dynamic lists of layouts that are sort-able and "design-able". 

The widget it self would be very similar to the list widget except instead of adding another of the same item, you get to choose which "layout" would work best for that specific content. Basically its a list that lets you choose predefined objects that contain other widgets.

Say I wanted to make a post. I have body copy, an image, a quote and gallery of images. Sure I could hard code all of those widgets to a post or use a the WYSIWYG (which will spit out what its going to spit out!), but wouldn't it be nicer If I could just add them on the fly for each post?

This will aid when creating templates for posts as we can loop through the flexible widget and wrap and know exactly what html its going to produce.

Example template (Go/Hugo):

```
// layouts/posts/single.html

<div>
  {{ range $content := .Params.FlexWidget }}
    {{ if $content.type === 'body' }}
      <div class="body-copy">
        {{ $content }}     
      </div>
    {{ end }}
    {{ if $content.type === 'fullWidthImage' }}
      <div class="full-width-image">
        {{ $content }}     
      </div>
    {{ end }}
    {{ if $content.type === 'anotherLayout' }}
      <div class="anotherLayout">
        {{ $content }}     
      </div>
    {{ end }}
  {{ end }}
</div>
```

I'm guessing this could take advantage of the `list` widget and tweak it to allow different objects to be used. Not sure though. Lemme know what y'all think. Cheers!

### PR DESCRIPTION
**Pull request by anthonysapp**
**- Summary**
The motivation for this pull request is so the CMS can support a front matter property containing a list of modular content or *blocks* to construct a custom page output.

**- Test plan**
Add "modular" as widget type in config.yml. This content type accepts a "types" list instead of "fields". Functionality is similar to a **list** widget with its valueType set to **valueTypes.SINGLE**. Added the functionality of a dropdown control at the top of the list which chooses the content type to be added to the modular content list. Because it is a variation of the List widget, it retains the ability to re-order child widgets via drag and drop.

configyml

*Modular content configuration*

modular-content-widget
*Modular content widget*

**- Description for the changelog**
Added modular Widget for saving modular content lists to frontmatter.

cute animal

fixes #1066

### CODE REVIEW

**Comment by verythorough** [comment]
Deploy preview for *netlify-cms-www* ready!

Built with commit 20c1ac5afa7c18eef7efaa57b25e341817bb441e

https://deploy-preview-1169--netlify-cms-www.netlify.com

**Comment by verythorough** [comment]
Deploy preview for *cms-demo* ready!

Built with commit 20c1ac5afa7c18eef7efaa57b25e341817bb441e

https://deploy-preview-1169--cms-demo.netlify.com

**Comment by anthonysapp** [comment]
@erquhart l will have a look tonight or tomorrow morning, for sure.

**Comment by Benaiah** [comment]
@erquhart What's the use case for your suggestion? Other widgets wrap the `List` widget for list functionality as well, and introducing this as a feature of the list widget, instead of a wrapper around it, seems kinda weird architecturally.

**Comment by anthonysapp** [comment]
@Benaiah I think maybe because it is a bit of code duplication? My widget is pretty close to the list widget and can probably be merged pretty easily.

**Comment by anthonysapp** [comment]
@erquhart I had a crack at this this morning. I think I've merged all the functionality in the appropriate places. Updated including moving documentation to the list component.

**Comment by erquhart** [comment]
@Benaiah what widget wraps the List widget? The PR as it was duplicated the entire list widget's code - I'm not sure how it could have imported the List widget and wrapped it to achieve this functionality.

The list widget already does 90% of this PR's goal - it just needs to allow more than one object definition to select from, versus always adding the one single object definition.

Architecturally, sharing list logic between widgets makes sense, but that can be a future abstraction. From a product standpoint, having the List widget be a bit more robust seems better than having two separate widgets.

**Comment by erquhart** [comment]
Hmm I am noticing one thing that I hadn't before - this isn't just selecting groups of fields, it also allows you to select a single field by name. The difference there is non-trivial.

**Comment by anthonysapp** [comment]
@erquhart I think this is why I decided against merging it into the List widget in the first place. Nevertheless it was a lot of code duplication so I think your reasoning had merit as well. In any case I merged into the list component and have all the functionality back as of the latest commit to the PR.

**Comment by erquhart** [comment]
@anthonysapp to help simplify the review, could you do the following?

* Re-add `dist` the `.gitignore` and remove that directory from the PR
* Remove any changes that only affect code style
* Remove the hardcoded previewing of specific widgets in the preview component

Let me know if you have questions on that last one.

**Comment by anthonysapp** [comment]
@erquhart Not sure I have any changes that only apply to code style? And yeah a pointer on that last one would help for sure. Thanks!

**Comment by verythorough** [comment]
Ooh, I am very much looking forward to this. I'm pretty sure I'll be able to use it to use Netlify CMS to edit the config.yml file, specifically to make it possible to add new collections via UI.

File: src/components/EditorWidgets/List/ListControl.js [hunk 1]
```diff
@@ -5,11 +5,15 @@ import { List, Map } from 'immutable';
...
+import { Icon, ListItemTopBar, Dropdown, DropdownItem } from 'UI';
 import ObjectControl from 'EditorWidgets/Object/ObjectControl';
 
 function ListItem(props) {
-  return <div {...props} className={`list-item ${ props.className || '' }`}>{props.children}</div>;
+  return (
+    <div {...props} className={`list-item ${ props.className || '' }`}>
+      {props.children}
+    </div>
+  );
```
Comment (erquhart) [inline src/components/EditorWidgets/List/ListControl.js]: Code style change.

File: src/components/EditorWidgets/List/ListControl.js [hunk 2]
```diff
@@ -23,23 +27,49 @@ function valueToString(value) {
...
+        </Dropdown>
+      </div>
+    )}
   </div>
 );
 
-const SortableList = SortableContainer(({ items, renderItem }) => {
-  return <div>{items.map(renderItem)}</div>;
-});
+const SortableList = SortableContainer(({ items, renderItem }) => <div>{items.map(renderItem)}</div>);
```
Comment (erquhart) [inline src/components/EditorWidgets/List/ListControl.js]: Code style change.

File: src/components/EditorWidgets/List/ListControl.js [hunk 3]
```diff
@@ -127,18 +162,28 @@ export default class ListControl extends Component {
...
     this.props.setActiveStyle();
-  }
+  };
 
   handleBlur = (e) => {
-    const listValue = e.target.value.split(',').map(el => el.trim()).filter(el => el);
+    const listValue = e.target.value
+      .split(',')
+      .map(el => el.trim())
+      .filter(el => el);
```
Comment (erquhart) [inline src/components/EditorWidgets/List/ListControl.js]: Code style change.

File: src/components/EditorWidgets/List/ListControl.js [hunk 4]
```diff
@@ -127,18 +162,28 @@ export default class ListControl extends Component {
...
     this.setState({ value: valueToString(listValue) });
     this.props.setInactiveStyle();
   };
 
   handleAdd = (e) => {
     e.preventDefault();
     const { value, onChange } = this.props;
-    const parsedValue = (this.valueType === valueTypes.SINGLE) ? null : Map();
+    const parsedValue = this.valueType === valueTypes.SINGLE ? null : Map();
```
Comment (erquhart) [inline src/components/EditorWidgets/List/ListControl.js]: This line is a code style change.

File: src/components/EditorWidgets/List/ListControl.js [hunk 5]
```diff
@@ -171,29 +223,36 @@ export default class ListControl extends Component {
     this.setState({ itemsCollapsed: itemsCollapsed.delete(index) });
 
     onChange(value.remove(index), parsedMetadata);
-  }
+  };
```
Comment (erquhart) [inline src/components/EditorWidgets/List/ListControl.js]: Code style change.

File: src/components/EditorWidgets/List/ListControl.js [hunk 6]
```diff
@@ -171,29 +223,36 @@ export default class ListControl extends Component {
...
 
   handleItemCollapseToggle = (index, event) => {
     event.preventDefault();
     const { itemsCollapsed } = this.state;
     const collapsed = itemsCollapsed.get(index);
     this.setState({ itemsCollapsed: itemsCollapsed.set(index, !collapsed) });
-  }
+  };
```
Comment (erquhart) [inline src/components/EditorWidgets/List/ListControl.js]: Code style change.

File: src/components/EditorWidgets/List/ListControl.js [hunk 7]
```diff
@@ -171,29 +223,36 @@ export default class ListControl extends Component {
...
 
   handleCollapseAllToggle = (e) => {
     e.preventDefault();
     const { value } = this.props;
     const { itemsCollapsed } = this.state;
     const allItemsCollapsed = itemsCollapsed.every(val => val === true);
     this.setState({ itemsCollapsed: List(Array(value.size).fill(!allItemsCollapsed)) });
-  }
+  };
```
Comment (erquhart) [inline src/components/EditorWidgets/List/ListControl.js]: Code style change.

File: src/components/EditorWidgets/List/ListControl.js [hunk 8]
```diff
@@ -226,44 +285,48 @@ export default class ListControl extends Component {
...
+          getAsset={getAsset}
+          onOpenMediaLibrary={onOpenMediaLibrary}
+          mediaPaths={mediaPaths}
+          onAddAsset={onAddAsset}
+          onRemoveInsertedMedia={onRemoveInsertedMedia}
+          classNameWrapper={`${ classNameWrapper } nc-listControl-objectControl`}
+          forList
+        />
+      </SortableListItem>
+    );
```
Comment (erquhart) [inline src/components/EditorWidgets/List/ListControl.js]: This return statement change is a code style change.

File: src/components/EditorWidgets/List/ListControl.js [hunk 9]
```diff
@@ -280,18 +343,20 @@ export default class ListControl extends Component {
...
+      <input
+        type="text"
+        id={forID}
+        value={value}
+        onChange={this.handleChange}
+        onFocus={this.handleFocus}
+        onBlur={this.handleBlur}
+        className={classNameWrapper}
+      />
+    );
```
Comment (erquhart) [inline src/components/EditorWidgets/List/ListControl.js]: This return statement change is a code style change.

File: src/components/EditorWidgets/List/ListControl.js [hunk 10]
```diff
@@ -280,18 +343,20 @@ export default class ListControl extends Component {
...
   }
-};
+}
```
Comment (erquhart) [inline src/components/EditorWidgets/List/ListControl.js]: Code style change.

File: src/components/EditorWidgets/Object/ObjectControl.js [hunk 11]
```diff
@@ -26,11 +26,7 @@ export default class ObjectControl extends Component {
     onAddAsset: PropTypes.func.isRequired,
     onRemoveInsertedMedia: PropTypes.func.isRequired,
     getAsset: PropTypes.func.isRequired,
-    value: PropTypes.oneOfType([
-      PropTypes.node,
-      PropTypes.object,
-      PropTypes.bool,
-    ]),
+    value: PropTypes.oneOfType([PropTypes.node, PropTypes.object, PropTypes.bool]),
```
Comment (erquhart) [inline src/components/EditorWidgets/Object/ObjectControl.js]: Code style change.

File: src/components/EditorWidgets/Object/ObjectControl.js [hunk 12]
```diff
@@ -90,19 +84,27 @@ export default class ObjectControl extends Component {
 
   handleCollapseToggle = () => {
     this.setState({ collapsed: !this.state.collapsed });
-  }
+  };
```
Comment (erquhart) [inline src/components/EditorWidgets/Object/ObjectControl.js]: Code style change.

File: src/components/EditorWidgets/Object/ObjectControl.js [hunk 13]
```diff
@@ -90,19 +84,27 @@ export default class ObjectControl extends Component {
...
+          {collapsed ? null : this.controlFor(value)}
+        </div>
+      );
+    } else if (multiFields) {
       return (
         <div id={forID} className={c(classNameWrapper, 'nc-objectControl-root')}>
-          { forList ? null : <TopBar collapsed={collapsed} onCollapseToggle={this.handleCollapseToggle} /> }
-          { collapsed ? null : multiFields.map((f, idx) => this.controlFor(f, idx)) }
+          {forList ? null : <TopBar collapsed={collapsed} onCollapseToggle={this.handleCollapseToggle} />}
+          {collapsed ? null : multiFields.map((f, idx) => this.controlFor(f, idx))}
```
Comment (erquhart) [inline src/components/EditorWidgets/Object/ObjectControl.js]: Code style change.

File: src/components/EditorWidgets/Object/ObjectPreview.js [hunk 14]
```diff
@@ -1,12 +1,116 @@
...
+  }
+  return null;
+};
+
+const renderValue = (field, value, getAsset) => {
+  const widget = value.get('widget');
+  let result = value.get('___name') || value.get(value.get('name'));
+  let valuePreview = null;
+
+  switch (widget) {
```
Comment (erquhart) [inline src/components/EditorWidgets/Object/ObjectPreview.js]: This switch statement ties this widget to the widgets below in an undesirable way - the object widget shouldn't know about these other widgets.

File: website/site/content/docs/widgets/list.md [hunk 15]
```diff
@@ -8,14 +8,19 @@ type: "widget"
```
Comment (erquhart) [inline website/site/content/docs/widgets/list.md]: The only changes this file should be additions - changes like switching the list item marker from `*` to `-` are code style changes.

**Comment by erquhart** [comment]
Cool, we'll have to get some eyes on that preview component and consider an alternate approach.

There are still some points from my review that need to be addressed - if you look at the file changes, there are a number of comments still showing. Let me know if you have questions on any of them. Mostly code style changes still remaining, but also `.gitignore` should have no changes in this PR (currently `dist` is still being added`), and the `dist` files that were added to the PR should be removed.

**Comment by anthonysapp** [comment]
@erquhart  I had fixed the .gitignore and removed the dist files before the last commit, so not sure why you were still seeing those. Apologies - I think I got to all the code style issues now.

File: src/components/EditorWidgets/List/ListControl.js [hunk 16]
```diff
@@ -143,6 +176,13 @@ export default class ListControl extends Component {
     onChange((value || List()).push(parsedValue));
   };
 
+  handleAddType = ({ itemType }) => {
+    const { value, onChange } = this.props;
+    const parsedValue = itemType;
+    this.setState({ itemsCollapsed: this.state.itemsCollapsed.push(false) });
+    onChange((value || List()).push(parsedValue));
```
Comment (michalmo) [inline src/components/EditorWidgets/List/ListControl.js]: one thing I noticed when trying to use this: the whole of the definition from types is written to the output and you end up with widget and label and the other configuration fields in your formatted file.

this is then read in render, and the widget rendering depends on the whole of the type definition being in the the value object.

ideally this would just store the type name and data in the output, and get the type definition from this.props.types based on the stored type name when rendering.

**Comment by erquhart** [comment]
Rebased and removed a few more code style changes.

**Comment by cjbrooks12** [comment]
I'm testing this widget out, and I found what I would consider to be a pretty serious issue. 

## tl;dr

**Problem**
The widget passes the full "type" config through to the Front Matter. When re-loading the widget from a file, it uses the fields config from the Front Matter to setup the widget, rather than looking up the appropriate config from the original `config.yml` definition.

**Solution**
Rendering the widget to Front Matter should pass along the actual field value, along with the "name" of the widget type the data represents. This field "name" should be sufficient to find the corresponding type from the original field list, from which the fields can be loaded.

**Additional widget option**
Allow customization of the "key" under which the widget type is saved. Right now, it saves it as `name`, but it would be nice to be able to consider a different property in the Front Matter to hold the "type" of the config, such as `type` or `kind`, being left to an option in the list field config 

### Full description

Given the following `config.yml`:
```yaml
...
collections:
  - name: "kitchenSink"
    label: "Kitchen Sink"
    folder: "_sink"
    create: true
    fields:
      - label: "Modular Content Chunks"
        label_singular: "Modular Content Chunk"
        name: "components_modular"
        widget: "list"
        types:
          - {label: "Readme", name: "readme", widget: "string"}
          - label: "Component Type One"
            name: "c1"
            widget: "object"
            fields:
              - {label: "Title", name: "title", widget: "string"}
      - label: "Non-modular list"
        label_singular: "Non-modular list"
        name: "components_list"
        widget: "list"
        fields:
          - {label: "Readme", name: "readme", widget: "string"}
      - {label: "Title", name: "title", widget: "string"}
      - {label: "Body", name: "body", widget: "markdown"}
```

Using the widget produces the following Front Matter entry and CMS display:

```yaml
components_modular:
  - label: Readme
    name: readme
    readme: Readme Value
    widget: string
  - c1:
      title: Component Type One Value
    fields:
      - label: Title
        name: title
        widget: string
    label: Component Type One
    name: c1
    widget: object
components_list:
  - readme: Readme Value
```



The data I need _is there_ in the front matter, but it's mixed in with a bunch of stuff that should stay in the `config.yml`. In fact, the widget seems to read data for the type's fields based on the Front Matter config, not `config.yml`, and as such, trimming the Front Matter data to what I would expect it to be, breaks the widget:

```yaml
components_modular:
  - name: readme
    readme: Readme Value
  - name: c1
    c1:
      title: Component Type One Value
components_list:
  - readme: Readme Value
```

**Comment by cjbrooks12** [comment]
Another usability suggestion while I'm thinking of it. The normal list widget allows items of either a single field widget or multiple, and it would be nice for the same type of functionality to be available from _each_ of the type widgets. 

Right now each item acts as the single field type, where the item type _is_ the field type, which also means that the field value is at a key that is always the same as the type itself. It might make more sense to have the list of types be configurations of "widget types", each with a "field" or "fields" property just as the non-modular list has in its root config.

In this way, we could almost consider the normal list widget to just be a special case of the modular list widget when there is only one type available, where it doesn't require a "name" property on the field value to be set.

**Comment by erquhart** [comment]
Been meaning to follow up on this - great review @cjbrooks12, we definitely only want the value being output, didn't notice the config was being passed.

I do think that simply using the name for the key should suffice, at least for the first iteration, and the built in dropdown already allows the type to be selected per item. Am I misunderstanding what you meant on that?

**Comment by cjbrooks12** [comment]
My thought with the name being configurable is that you could make the widget understand both the following snippets, so the widget can be adapted to whatever terminology the developer prefers or SSG requires.

```yaml
components_modular:
  - name: readme
    readme: Readme Value
  - name: c1
    c1:
      title: Component Type One Value
```

```yaml
components_modular:
  - type: readme
    readme: Readme Value
  - type: c1
    c1:
      title: Component Type One Value
```

The change from "name" to "type" would be controlled from the `config.yml` widget config like so:

```yaml
collections:
  - name: "kitchenSink"
    label: "Kitchen Sink"
    folder: "_sink"
    create: true
    fields:
      - label: "Modular Content Chunks"
        label_singular: "Modular Content Chunk"
        name: "components_modular"
        widget: "list"
        typeKey: 'type' # defaults to 'name'
        types: ...
```

**Comment by erquhart** [comment]
Oh duh we need a way to map values back to types on read 🤦‍♂️ 

I'm with you now, agreed.

**Comment by morrislaptop** [comment]
Can't wait for this!

**Comment by erquhart** [comment]
Hey @anthonysapp just checking in - still planning to circle back on the issues raised in the latest comments?

**Comment by anthonysapp** [comment]
Hey all this is some great discussion. So sorry to have been unresponsive. I'm quite busy at the moment on another project but I would indeed like to loop back to this in the next week or so, as I'd love to use it for an upcoming project myself :).
If anyone else feels the urge to help out, by all means.

**Comment by shaunbent** [comment]
Awesome work on this @anthonysapp - I need this exact feature right now. I've sort of managed to recreate it using `registerEditorComponent` but its not the best user experience and it also means might front end is a little heavier as I need to transform my Markdown AST to React on the client.

This PR/feature would completely solve my problem. I really hope this makes it in soon.

**Comment by anthonysapp** [comment]
@erquhart @cjbrooks12 hi guys, I took a look at this last night and I'm just curious if you both think this should still be merged with the List widget, or you think it warrants its own widget with the proposed changes? Also - I'm not 100% sure I understand this [comment](#issuecomment-382531603). What is inherently wrong with using the "name" field to reference each component? I know I'm probably missing something easy :)

**Comment by cjbrooks12** [comment]
Technically-speaking, I would see the List widget as a special case of the Modular widget where there is only one available type. I think there is still a case for keeping the Modular widget separate for clarity of intent. 

As for the practical reason for wanting to customize the field used as the key, it is simply that the consumers of these modular blocks may consume them differently than Netlify CMS does. 

For example, I'm build a SSG with this kind of modularity baked into the framework itself, and Orchid reads these blocks from the `type` field, not `name`. [See here](https://github.com/JavaEden/Orchid/edit/dev/OrchidCore/src/orchid/resources/homepage.md) for an example: the `components` top-level field is the modular container, and each list item is a modular block identified by its `type`. Orchid wouldn't be able to understand the output of this widget if it only worked with `name`.

Or take [Forestry's Blocks](https://forestry.io/docs/settings/fields/blocks/), which uses `template` to identify the type of the modular item. If someone were to migrate to Netlify CMS from Forestry, wouldn't it be easier to tell the Netlify widget to look for the `template` parameter rather than finding and changing all instances of `template` with `name` and hoping you got it all right? 

As far as making it happen in the code, it shouldn't be much of a change. Rather than doing 

```
item.name
```

to get the modular type, it would just be one step back with a configuration value to get it instead:

```
// config.typeKey defaults to 'name' but is configurable by the end-user
item[config.typeKey]
```

**Comment by anthonysapp** [comment]
@cjbrooks12 I think I'm following now. Thanks for the feedback!

**Comment by gil--** [comment]
Just thought I'd say that this is an extremely exciting widget and would really elevate the possibilities (custom layouts on a per page basis). 

Looking forward to seeing it shipped.

**Comment by nerikh** [comment]
FYI - I'm monitoring the status of this modular content widget as it is all I'm waiting for to jump into Netlify CMS.

**Comment by erquhart** [comment]
Closing as stale - if someone wants to pick it up, please comment.

**Comment by airtonix** [comment]
it would be great if we could use a group of other collections in place of `types`. This way, existing blocks are re-usable.

DjangoCMS and EPiServer calls these blocks.

And if you have a collection of blocks that are actually layout types with further modular widgets in them, you can do awesome things.

**Comment by criticalmash** [comment]
What will it take to get this past the finish line? Are there particular features that still need to be implemented? or tests to pass?

I'm willing to help out, but I have zero React experience.

**Comment by loremipson** [comment]
Hey, our team is currently exploring some ideas that are very similar on our own fork. We'll give an update when we feel more confident in our approach here.

**Comment by mittalyashu** [comment]
> As discussed in the netlifyCMS community meeting.

I was trying to say this that it can split the `.json` file _(each object)_ into different card.

**Comment by alexanderfountain** [comment]
This sounds awesome. I have been unable to get this pull working at all with current version of netlifycms though.

**Comment by chan360** [comment]
This sounds like a must have widget when it comes to customizability and I'm wondering if there has been new updates on this?

**Comment by erquhart** [comment]
No updates, still needs a sponsor.

**Comment by erquhart** [comment]
For those watching this issue, a new implementation just popped up courtesy of @smnh in #1857 - we need some reviewing and testing, please check it out and make comments!
